lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,499 @@
+  CircuitOp circuitOp = getOperation();
+
+  AnnotationSet annotations(circuitOp);
+  if (annotations.empty())
+    return;
+
+  auto context = &getContext();
+  OpBuilder builder(context);
+  builder.setInsertionPointToEnd(getOperation()->getBlock());
+
+  // Store a mapping of interface name to InterfaceOp.
+  llvm::StringMap<sv::InterfaceOp> interfaces;
+
+  // Track the order that interfaces should be emitted in.
+  llvm::SmallVector<std::pair<StringRef, StringRef>> interfaceKeys;
+
+  // Examine the Circuit's Annotations doing work to remove Grand Central
+  // Annotations.  Ignore any unprocesssed annotations and rewrite the Circuit's
+  // Annotations with these when done.
+  llvm::SmallVector<Attribute> unprocessedAnnos;
+  for (auto anno : annotations) {
+    if (!anno.isClass("sifive.enterprise.grandcentral.AugmentedBundleType")) {
+      unprocessedAnnos.push_back(anno.getDict());
+      continue;
+    }
+
+    AugmentedBundleType bundle;
+    if (auto maybeBundle = decodeBundleType(anno))
+      bundle = maybeBundle.getValue();
+    else {
+      emitError(circuitOp.getLoc(),
+                "'firrtl.circuit' op contained an 'AugmentedBundleType' "
+                "Annotation which did not conform to the expected format")
+              .attachNote()
+          << "the problematic 'AugmentedBundleType' is: '" << anno.getDict()
+          << "'";
+      return signalPassFailure();
+    }
+
+    for (auto elt : bundle.elements) {
+      std::pair<StringRef, StringRef> key = {bundle.defName, elt.name};
+      interfaceMap[key] = {elt.tpe, elt.description};
+      interfaceKeys.push_back(key);
+    }
+
+    // If the interface already exists, don't create it.
+    if (interfaces.count(bundle.defName))
+      continue;
+
+    // Create the interface.  This will be populated later.
+    interfaces[bundle.defName] =
+        builder.create<sv::InterfaceOp>(circuitOp->getLoc(), bundle.defName);
+  }
+
+  circuitOp->setAttr("annotations", builder.getArrayAttr(unprocessedAnnos));
+
+  // Walk through the circuit to collect additional information.  If this fails,
+  // signal pass failure.
+  for (auto &op : circuitOp.getBody()->getOperations()) {
+    if (isa<FModuleOp, FExtModuleOp>(op)) {
+      GrandCentralVisitor visitor(&getContext(), &interfaceMap);
+      visitor.visitModule(&op);
+      if (visitor.hasFailed())
+        return signalPassFailure();
+    }
+  }
+
+  // Populate interfaces.
+  for (auto a : interfaceKeys) {
+    auto defName = a.first;
+    auto name = a.second;
+
+    auto info = &interfaceMap[{defName, name}];
+    if (info->isMissing()) {
+      emitError(circuitOp.getLoc())
+          << "'firrtl.circuit' op contained a Grand Central Interface with "
+             "defName '"
+          << defName << "' that had a component called '" << name
+          << "' which did not have a scattered companion annotation.  This may "
+             "indicate that the companion was improperly deleted.";
```

Comment:
Yeah, this is currently the error path for a user specifying incorrect annotations. A user can hit this if they have an incorrect target as part of an element. 

That said, the only way a user can currently hit this is to manually write the annotation file (or construct the annotations manually in Chisel) and use element targets that point to things that don't exist in the circuit. If the user relies on the APIs for Grand Central in Chisel, then this should be impossible to do.

This could also be a compiler bug, e.g., LowerTypes is destroying annotations during lowering and then you can wind up here.

I think the expected path, for now, is the user path only. I'll switch this to remove the latter part about improper deletion. Once a solution to Annotation Verification (https://github.com/llvm/circt/issues/1013) comes online, this may be able to turned off.